DockerLogsContainer implementation

double16 opened the pull request:
This task implements `docker logs`. Fixes #132

cdancy commented:
@double16 thanks for the PR! A few things:

1. Any particular reason you created a separate functional test file? We've been putting all functional tests inside `DockerWorkFlowFunctionalTest`.
2. For loading classes from docker-java please follow the model we set within `ThreadContextClassLoader` and it's impl `DockerThreadContextClassLoader`.
3. Consider creating an inner class for the `tail` variables. Having 1 variable mean potentially N number of things will be confusing for users. Take a look at `DockerCreateContainer` to see how we've done similar things in the past.

cdancy commented on src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerLogsContainer.groovy:
```diff
@@ -0,0 +1,92 @@
+            logsCommand.withTimestamps(showTimestamps)
+        }
+
+        logsCommand.withStdOut(true).withStdErr(true)
```
Should make these variables and default to true. If it's an option then it should be exposed to users to toggle whichever they may need.

double16 commented:
I'll address these and one more thing ... I forgot to add documentation. :blush:

cdancy commented:
On further thinking: lets keep the functional tests where you have them now. Our all-in-one functional test class is getting quite big and I think making it bigger, if only to keep tests in one place, is not the way to go.

double16 commented:
I've addressed your comments. For the tail, I opted to go with two tail options similar to the docker-java API and throw an exception if both are specified.

cdancy commented on src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerLogsContainer.groovy:
```diff
@@ -0,0 +1,150 @@
+        logsCommand.withStdOut(getStdOut()).withStdErr(getStdErr())
+
+        if (getTailAll() != null && getTailCount() != null) {
+            throw new InvalidUserDataException("Conflicting parameters: only one of tailAll and tailCount can be specified")
```
Throw GradleException instead. It's standard to expect as much from a gradle plugin.

double16 replied:
InvalidUserDataException extends GradleException, so being more specific isn't hurting handling the more generic case, right? According to Gradle API, InvalidUserDataException is appropriate here:
```
A InvalidUserDataException is thrown, if a user is providing illegal data for the build.
```

cdancy replied:
Ahhhh gotcha. I assumed it was java exception. Ok lets keep it. The standard we've been keeping thus far is to use the generic GradleException in all places. Having something a bit more specific can't hurt.

cdancy commented on src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerLogsContainer.groovy:
```diff
@@ -0,0 +1,150 @@
+    @Optional
+    Date since
+
+    private createLoggingCallback() {
```
Stick this method into `DockerThreadContextClassLoader` and add its interface definition to `ThreadContextClassLoader`. Look at any of the other tasks to see how we do this.

cdancy commented on src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerLogsContainer.groovy:
```diff
@@ -0,0 +1,150 @@
+
+    @Override
+    void runRemoteCommand(dockerClient) {
+        logger.info "Logs for container with ID '${getContainerId()}'."
```
Lets use `logger.quiet` here.

cdancy commented on src/main/groovy/com/bmuschko/gradle/docker/tasks/container/DockerLogsContainer.groovy:
```diff
@@ -0,0 +1,150 @@
+
+            invoke: {Object proxy, java.lang.reflect.Method method, Object[] args ->
+                if ("onNext" == method.name) {
+                  def frame = args[0]
```
Are we sure `frame` wont be null? We had an issue before within [createPrintStreamProxyCallback](https://github.com/bmuschko/gradle-docker-plugin/blob/master/src/main/groovy/com/bmuschko/gradle/docker/utils/DockerThreadContextClassLoader.groovy#L350) where the returned 'stream' was actually null causing failures for users.

cdancy commented on src/main/groovy/com/bmuschko/gradle/docker/utils/DockerThreadContextClassLoader.groovy:
```diff
@@ -339,6 +339,40 @@ class DockerThreadContextClassLoader implements ThreadContextClassLoader {
+                  switch (frame.streamType as String) {
+                    case "STDOUT":
+                    case "RAW":
+                        System.out.print(new String(frame.payload))
```
lets change this to logger.quiet

double16 replied:
That means normal output is only shown when Gradle is run with `--info` or `--debug`. From my point of view, if I'm using `DockerLogsContainer` I intend to see container output as a part of the normal Gradle run. You sure you want this to be logger output?

cdancy replied:
logger.quiet will be shown by default. logger.info will only be shown if the `-i` flag is given. This is why we have a simple logger.quiet message for each task to give some indication something has happened. Is this not happening in your case?

double16 replied:
Sorry, I was confused. logger.quiet is showing up for me by default. I've made the requested logging changes.

cdancy commented on src/main/groovy/com/bmuschko/gradle/docker/utils/DockerThreadContextClassLoader.groovy:
```diff
@@ -339,6 +339,40 @@ class DockerThreadContextClassLoader implements ThreadContextClassLoader {
+                        System.out.print(new String(frame.payload))
+                        break
+                    case "STDERR":
+                        System.err.print(new String(frame.payload))
```
and this to logger.error

double16 replied:
Following the logic for standard out, if it continues to go to stdout, for consistency I think this should go to stderr. If standard out goes to the logger, then logger.error makes sense here

cdancy commented:
@double16 just a couple minor things related to logging and I think we are set. Have you run the integration tests yet?

double16 commented:
@cdancy I have run all of the tests for the project and they pass, except for three that push to DockerHub and private repos, which I didn't configure.

cdancy commented:
@double16 lgtm. Squash and rebase then I'll merge.

double16 commented:
@cdancy squashed and ready to go

cdancy commented:
Thanks @double16!

cdancy commented:
@double16 have you used this task at all? I'm more wondering how you are using it assuming you're doing so. My thoughts surround the capturing of the logs themselves and if you're doing so through some mechanism. I'm thinking of putting together some sort of `sink` for this task or some way to capture the output so folks can process it and/or make some decision further downstream. With all that in mind I'm wondering how you are going about things and if you have any thoughts surrounding such a thing.

double16 commented:
I am currently using this task in a Bamboo plan. Bamboo captures the Gradle output and saves the logs so if something goes wrong (or it succeeds), the logs can be inspected. I've really only used the logs to review when something goes wrong.

I avoid making decisions based on log output. I rely on output files with a specific purpose such as JUnit XML, or JSON, etc.

Allowing a sink to be specified sounds like a good idea, as long as it's optional. I'd really like to keep the functionality to output to stdout/stderr so CI tools like Bamboo or Jenkins can capture the output.

cdancy commented:
Aaahhhh a fellow Bamboo user! Our use case here is more for checking when a given container is "live" by waiting for our app to print out an "init is complete" message.

I'd definitely keep the functionality as-is, as we've already released and folks will have come to depend on it. Any additions would be purely optional at this point.

Thanks for the info.
